src/era_validator.rs

```diff
@@ -1,143 +1,163 @@
-                    log::info!("syncing new epoch: {}", epoch);
-                }
-                Err(e) => {
-                    return {
-                        log::error!("error: {}", e);
-                        Err(EraValidateError::EpochAccumulatorError)
+impl EraValidator for PreMergeAccumulator {
+    type Error = HeaderAccumulatorError;
+
+    fn era_validate(
+        &self,
+        mut headers: Vec<ExtHeaderRecord>,
+        start_epoch: usize,
+        end_epoch: Option<usize>,
+        use_lock: bool,
+    ) -> Result<Vec<usize>, Self::Error> {
+        let end_epoch = end_epoch.unwrap_or(start_epoch + 1);
+
+        // Ensure start epoch is less than end epoch
+        if start_epoch >= end_epoch {
+            Err(EraValidateError::EndEpochLessThanStartEpoch)?;
+        }
+
+        let mut validated_epochs = Vec::new();
+        for epoch in start_epoch..end_epoch {
+            // checks if epoch was already synced form lockfile.
+            if use_lock {
+                let file_path = Path::new("./lockfile.json");
+                let lock_file = Lock::from_file(file_path)?;
+
+                match lock_file.check_sync_state(file_path, epoch, self.historical_epochs[epoch].0)
+                {
+                    Ok(true) => {
+                        log::info!("Skipping, epoch already synced: {}", epoch);
+                        continue;
+                    }
+                    Ok(false) => {
+                        log::info!("syncing new epoch: {}", epoch);
+                    }
+                    Err(e) => {
+                        return {
+                            log::error!("error: {}", e);
+                            Err(EraValidateError::EpochAccumulatorError.into())
+                        }
                     }
                 }
             }
-        }
-        let epoch_headers: Vec<ExtHeaderRecord> = headers.drain(0..MAX_EPOCH_SIZE).collect();
-        let root = process_headers(epoch_headers, epoch, &macc)?;
-        validated_epochs.push(epoch);
-        // stores the validated epoch into lockfile to avoid validating again and keeping a concise state
-        if use_lock {
-            match store_last_state(Path::new("./lockfile.json"), LockEntry::new(&epoch, root)) {
-                Ok(_) => {}
-                Err(e) => {
-                    log::error!("error: {}", e);
-                    return Err(EraValidateError::EpochAccumulatorError);
+            let epoch_headers: Vec<ExtHeaderRecord> = headers.drain(0..MAX_EPOCH_SIZE).collect();
+            let root = self.process_headers(epoch_headers, epoch)?;
+            validated_epochs.push(epoch);
+
+            // stores the validated epoch into lockfile to avoid validating again and keeping a concise state
+            if use_lock {
+                let path = Path::new("./lockfile.json");
+                let mut lock_file = Lock::from_file(path)?;
+                lock_file.update(LockEntry::new(&epoch, root));
```

Comment:
This is good, update the entries in memory and then add it to the file. I was wondering about this, will we have significant performance issues with this? Instead of just appending to the file. Because in line 105, it uses the truncate option, which sets the length of the file to be 0.
I'm concerned about performance later because there can be a lot of writes into this file, or the file can be somewhat large, not sure if significant.

Reply:
Don't prematurely optimize. If this becomes an issue we benchmark and then solve it when there's a use case for doing so.